src/adapter/src/coord.rs

```diff
@@ -6761,6 +6784,94 @@ mod timeline {
             self.advance_to.take()
         }
     }
+
+    /// Interval used to persist durable timestamps. See [`DurableTimestampOracle`] for more
+    /// details.
+    pub static TIMESTAMP_PERSIST_INTERVAL: Lazy<mz_repr::Timestamp> = Lazy::new(|| {
+        Duration::from_secs(15)
+            .as_millis()
+            .try_into()
+            .expect("15 seconds can fit into `Timestamp`")
+    });
+
+    /// A type that wraps a [`TimestampOracle`] and provides durable timestamps. This allows us to
+    /// recover a timestamp that is larger than all previous timestamps on restart. The protocol
+    /// is based on timestamp recovery from Percolator <https://research.google/pubs/pub36726/>. We
+    /// "pre-allocate" a group of timestamps at once, and only durably store the largest of those
+    /// timestamps. All timestamps within that interval can be served directly from memory, without
+    /// going to disk. On restart, we re-initialize the current timestamp to a value one larger
+    /// than the persisted timestamp.
+    ///
+    /// See [`TimestampOracle`] for more details on the properties of the timestamps.
+    pub struct DurableTimestampOracle<T> {
+        timestamp_oracle: TimestampOracle<T>,
+        durable_timestamp: T,
+        persist_interval: T,
+    }
+
+    impl<T: super::CoordTimestamp> DurableTimestampOracle<T> {
+        /// Create a new durable timeline, starting at the indicated time. Timestamps will be
+        /// allocated in groups of size `persist_interval`.
+        ///
+        /// See [`TimestampOracle::new`] for more details.
+        pub fn new<F>(initially: T, next: F, persist_interval: T) -> Self
+        where
+            F: Fn() -> T + 'static,
+        {
+            Self {
+                timestamp_oracle: TimestampOracle::new(initially.clone(), next),
+                durable_timestamp: initially,
+                persist_interval,
+            }
+        }
+
+        /// Peek the current value of the timestamp.
+        ///
+        /// See [`TimestampOracle::peek_ts`] for more details.
+        pub fn peek_ts(&self) -> T {
+            self.timestamp_oracle.peek_ts()
+        }
+
+        /// Acquire a new timestamp for writing. Optionally returns a timestamp that needs to be
+        /// persisted to disk.
+        ///
+        /// See [`TimestampOracle::write_ts`] for more details.
+        pub fn write_ts(&mut self) -> (T, Option<T>) {
+            let ts = self.timestamp_oracle.write_ts();
+            self.maybe_allocate_new_timestamps(ts)
+        }
+
+        /// Acquire a new timestamp for reading.
+        ///
+        /// See [`TimestampOracle::read_ts`] for more details.
+        pub fn read_ts(&mut self) -> T {
+            self.timestamp_oracle.read_ts()
+        }
```

Comment:
Getting a read timestamp never increments the global timestamp, only write timestamps do that.

Reply:
I see. That's now a new requirement of `TimestampOracle`, though? It has a `next: Box<dyn Fn() -> T>`, but with this change it would be an error to use that in its `read_ts` method. I wonder if we can thread that information though?

Reply:
Yeah `read_ts` does not call `next`. It does potentially update `advance_to`, which is why it needs to be `mut`. I'm not sure how we can thread that info through except just adding a comment to `read_ts` explaining this.

Reply:
I added the defensive logic

Reply:
Actually just added an assert, adding the defensive logic means that the function needs to be `aync` which is really annoying because that propagates to a bunch of other methods.